### PR TITLE
Rename MagicFunctionArg to AugmentedFunction + docstrings

### PR DESCRIPTION
**Pull request by jni**
# Description

- rename new typedef
- update docstrings to be more descriptive.

Note: currently, the autogenerated documentation for `add_dock_widget` and `add_function_widget` lives on the `Window` object for reasons that I don't quite understand. I can update the links when we fix that.

## Type of change

- [ ] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: napari/plugins/hook_specifications.py [hunk 1]
```diff
@@ -331,26 +331,36 @@ def napari_write_vectors(path: str, data: Any, meta: dict) -> Optional[str]:
...
+    function(s) : callable, tuple of callable & dict(s), or list thereof
+        Implementations should provide either a single function, or a list of
+        functions. The functions should have Python type annotations so that
+        `magicgui <https://napari.org/magicgui>`_ can generate a widget from
+        them. Each function can be provided as-is or as part of a 2-tuple or 3-tuple
+        with configuration dicts. The second element in the tuple should be a
+        dictionary defining magicgui `configuration options
+        <https://napari.org/magicgui/usage/configuration.html#magicgui-options>`_,
+        while the third element should provide keyword arguments for
+        :meth:`napari.qt.Window.add_dock_widget`.
```
Comment (sofroniewn) [inline napari/plugins/hook_specifications.py]: Add  "excluding `shortcut`, which is not yet supported."

File: napari/plugins/hook_specifications.py [hunk 2]
```diff
@@ -331,26 +331,36 @@ def napari_write_vectors(path: str, data: Any, meta: dict) -> Optional[str]:
...
 
     Returns
     -------
-    dock_widgets : QWidget class or list of QWidget classes
-        QWidget or tuple QWidget, dict.
+    dock_widget(s) : QWidget class or list of QWidget classes
+        Implementations should return either QWidget classes (one or a list),
+        or tuple(s) containing QWidget classes as well as a dictionary
+        containing keyword arguments for
+        :meth:`napari.qt.Window.add_dock_widget`.
```
Comment (sofroniewn) [inline napari/plugins/hook_specifications.py]: Add  "excluding `shortcut`, which is not yet supported."

**Review by sofroniewn — APPROVED** [review]
Made two suggestions, we don't let plugin provide their own shortcuts here due to worries about name space collisions. We should note that in the docks.

> Note: currently, the autogenerated documentation for add_dock_widget and add_function_widget lives on the Window object for reasons that I don't quite understand. I can update the links when we fix that.

What do you mean by this, i don't understand? those methods live on `viewer.window` and so the docs live there? should we move the methods?

Otherwise this looks good to me, and I'm approving. Curious if @tlambert03 has any additional comments before merge

**Comment by jni** [comment]
> What do you mean by this, i don't understand? those methods live on viewer.window and so the docs live there? should we move the methods?

Oh, sorry, I must have missed a discussion — since there was discussion of deprecating the public `.window`, I thought we wouldn't add to it. If that's its rightful place, all is well!

**Comment by sofroniewn** [comment]
> Oh, sorry, I must have missed a discussion — since there was discussion of deprecating the public .window, I thought we wouldn't add to it. If that's its rightful place, all is well!

Ah yes, you're right. This was being discussed, but hasn't happened yet!

**Pull request merged**